frame/bags-list/src/list/mod.rs

```diff
@@ -384,6 +384,73 @@ impl<T: Config> List<T> {
 		})
 	}
 
+	/// Put `heavier_id` to the position directly in front of `lighter_id`. Both ids must be in
+	/// same bag  and the `weight_of` `lighter_id` must be less than that of `heavier_id`.
+	pub(crate) fn put_in_front_of(
+		lighter_id: &T::AccountId,
+		heavier_id: &T::AccountId,
+		weight_of: Box<dyn Fn(&T::AccountId) -> VoteWeight>,
+	) -> Result<(), crate::pallet::Error<T>> {
+		use crate::pallet;
+		use frame_support::ensure;
+
+		let lighter_node = Node::<T>::get(&lighter_id).ok_or(pallet::Error::IdNotFound)?;
+		let mut heavier_node = Node::<T>::get(&heavier_id).ok_or(pallet::Error::IdNotFound)?;
+
+		ensure!(lighter_node.bag_upper == heavier_node.bag_upper, pallet::Error::NotInSameBag);
+
+		// this is the most expensive check, so we do it last.
+		ensure!(weight_of(&heavier_id) > weight_of(&lighter_id), pallet::Error::NotHeavier);
+
+		// we need to write the bag to storage if its head and/or tail is updated.
+		if lighter_node.is_terminal() || heavier_node.is_terminal() {
```

Comment:
Here you are assuming that the lighter node could only ever be a head, and the heavier node could only ever be a tail, while I don't see that being guaranteed. Can you elaborate all the edge cases that you considered, and why you think this is enough?

Reply:
here's an idea to break this down into smaller bite-size pieces: 

1. For removing, we simply use the current `fn remove()`. After all the checks, we immediately remove `heavier`. The logic for this function should already update the Bag head/tail if `heavier.is_terminal()`.
2. For inserting, we write a private helper function as such: 

```
fn remove_at(before: Option<>)
``` 

I think this logic will be easier to verify than the current code. 

If we keep the current code, here's a mental model to check it

1. none are terminal
2. lighter is head, heavier is not terminal (common case)
3. lighter is tail, heavier is not terminal (mistake case)
4. heavier is tail, lighter is not terminal (common case)
5. heavier is head, lighter is not terminal (mistake case)
6. lighter is head, heavier is tail (common case)
7. lighter is tail, heavier is head (mistake case)

I'm going to now start thinking about each case and how it is handled.

Reply:
My analysis shows case 5 is not handled correctly, when the non-terminal lighter is not adjacent to the heavier one. 

I'd like to see the fix, but my gut feeling is that rewriting it using smaller functions is easier. This `if conditions_in_which_we_might_update_bags {}` is a bit hard to verify. If we first remove, then update everything, and then insert, and update everything again, there would  probably be a few extra (IMO negligible) lookups to the overlay, but the (non-negligible) db access would be the same.

Reply:
fa1a33d3fa9fdc3a288413ea388d1239ad8f0b7d